File: BookJukBookJuk_WEB/src/lib/gestureClassifiers.ts
/** MediaPipe 손 랜드마크 규칙 기반 제스처 분류 (Python gesture_classifiers.py 포트). */

export type HandLandmark = { x: number; y: number; z?: number }

export type GestureId =
  | 'stop'
  | 'follow_me'
  | 'lead_again'
  | 'thumbs_up'
  | 'thumbs_down'
  | 'ok_sign'

const EXTEND_RATIO = 1.08
const THUMB_EXTEND_RATIO = 1.05
const OK_TOUCH_RATIO = 0.55
// Normalized landmark bbox area. Keeps distant/background hands from driving gestures.
export const CLOSEST_HAND_MIN_BOX_AREA = 0.045
type FingerState = Record<'thumb' | 'index' | 'middle' | 'ring' | 'pinky', boolean>

function dist(a: HandLandmark, b: HandLandmark): number {
  return Math.hypot(a.x - b.x, a.y - b.y)
}

function fingerExtended(lm: HandLandmark[], tipId: number, pipId: number): boolean {
  const w = lm[0]
  const tip = lm[tipId]
  const pip = lm[pipId]
  return dist(w, tip) > dist(w, pip) * EXTEND_RATIO
}

function thumbExtended(lm: HandLandmark[]): boolean {
  const w = lm[0]
  const tip = lm[4]
  const ip = lm[3]
  return dist(w, tip) > dist(w, ip) * THUMB_EXTEND_RATIO
}

function allFingers(lm: HandLandmark[]): FingerState {
  return {
    thumb: thumbExtended(lm),
    index: fingerExtended(lm, 8, 6),
    middle: fingerExtended(lm, 12, 10),
    ring: fingerExtended(lm, 16, 14),
    pinky: fingerExtended(lm, 20, 18),
  }
}

function isOpenPalm(f: FingerState): boolean {
  return Object.values(f).every(Boolean)
}

function isOkSign(lm: HandLandmark[], f: FingerState): boolean {
  const touch = dist(lm[4], lm[8]) < dist(lm[5], lm[17]) * OK_TOUCH_RATIO
  return touch && f.middle && f.ring && f.pinky
}

function isThumbPoseBase(f: FingerState): boolean {
  return f.thumb && !f.index && !f.middle && !f.ring && !f.pinky
}

function isThumbsUp(lm: HandLandmark[], f: FingerState): boolean {
  if (!isThumbPoseBase(f)) return false
  return lm[4].y < lm[3].y && lm[4].y < lm[0].y
}

function isThumbsDown(lm: HandLandmark[], f: FingerState): boolean {
  if (!isThumbPoseBase(f)) return false
  return lm[4].y > lm[3].y && lm[4].y > lm[0].y
}

function isLeadAgain(lm: HandLandmark[], f: FingerState): boolean {
  if (!(f.index && f.thumb)) return false
  if (f.middle || f.ring || f.pinky) return false
  const touch = dist(lm[4], lm[8]) < dist(lm[5], lm[17]) * OK_TOUCH_RATIO
  return !touch
}

function isFist(f: FingerState): boolean {
  return !Object.values(f).some(Boolean)
}

export function classifyOneHandGesture(lm: HandLandmark[]): GestureId | null {
  const fingers = allFingers(lm)
  if (isOpenPalm(fingers)) return 'stop'
  if (isThumbsUp(lm, fingers)) return 'thumbs_up'
  if (isThumbsDown(lm, fingers)) return 'thumbs_down'
  if (isOkSign(lm, fingers)) return 'ok_sign'
  if (isLeadAgain(lm, fingers)) return 'lead_again'
  if (isFist(fingers)) return 'follow_me'
  return null
}

export function handLandmarkBoxArea(lm: HandLandmark[]): number {
  if (lm.length === 0) return 0
  let minX = Number.POSITIVE_INFINITY
  let maxX = Number.NEGATIVE_INFINITY
  let minY = Number.POSITIVE_INFINITY
  let maxY = Number.NEGATIVE_INFINITY

  for (const point of lm) {
    minX = Math.min(minX, point.x)
    maxX = Math.max(maxX, point.x)
    minY = Math.min(minY, point.y)
    maxY = Math.max(maxY, point.y)
  }

  return Math.max(0, maxX - minX) * Math.max(0, maxY - minY)
}

export function pickClosestHandLandmarks(
  candidates: HandLandmark[][],
  minBoxArea = CLOSEST_HAND_MIN_BOX_AREA,
): HandLandmark[] | null {
  let closest: HandLandmark[] | null = null
  let largestArea = 0

  for (const lm of candidates) {
    const area = handLandmarkBoxArea(lm)
    if (area > largestArea) {
      largestArea = area
      closest = lm
    }
  }

  return closest && largestArea >= minBoxArea ? closest : null
}

export const GESTURE_CONFIRM_FRAMES = 8
export const GESTURE_COOLDOWN_FRAMES = 24

export const GESTURE_LABELS_KO: Record<GestureId, string> = {
  stop: '정지 (손 펼침)',
  follow_me: '나 따라와 (주먹)',
  lead_again: '다시 리드 (검지+엄지 ㄴ)',
  thumbs_up: '엄지 올림 (담기)',
  thumbs_down: '엄지 내림 (빼기)',
  ok_sign: 'OK 사인',
}

File: BookJukBookJuk_WEB/book_recognition/gesture_classifiers.py
"""MediaPipe 손 랜드마크 규칙 기반 제스처 분류 (mediapipe/cv2 의존 없음)."""

from __future__ import annotations

import math
from typing import Any, List, Optional

EXTEND_RATIO = 1.08
THUMB_EXTEND_RATIO = 1.05
OK_TOUCH_RATIO = 0.55

MOBILITY_GESTURES: dict[str, dict[str, str]] = {
    "stop": {"type": "command", "action": "stop"},
    "follow_me": {"type": "command", "action": "set_mode", "mode": "guidance"},
    "lead_again": {"type": "command", "action": "set_mode", "mode": "escort"},
}


def _dist(a: Any, b: Any) -> float:
    return math.hypot(a.x - b.x, a.y - b.y)


def finger_extended(lm: List[Any], tip_id: int, pip_id: int) -> bool:
    w, tip, pip = lm[0], lm[tip_id], lm[pip_id]
    return _dist(w, tip) > _dist(w, pip) * EXTEND_RATIO


def thumb_extended(lm: List[Any]) -> bool:
    w, tip, ip = lm[0], lm[4], lm[3]
    return _dist(w, tip) > _dist(w, ip) * THUMB_EXTEND_RATIO


def _all_fingers(lm: List[Any]) -> dict[str, bool]:
    return {
        "thumb": thumb_extended(lm),
        "index": finger_extended(lm, 8, 6),
        "middle": finger_extended(lm, 12, 10),
        "ring": finger_extended(lm, 16, 14),
        "pinky": finger_extended(lm, 20, 18),
    }


def _is_open_palm(lm: List[Any], fingers: Optional[dict[str, bool]] = None) -> bool:
    f = fingers if fingers is not None else _all_fingers(lm)
    return all(f.values())


def _is_ok_sign(lm: List[Any], fingers: Optional[dict[str, bool]] = None) -> bool:
    f = fingers if fingers is not None else _all_fingers(lm)
    touch = _dist(lm[4], lm[8]) < _dist(lm[5], lm[17]) * OK_TOUCH_RATIO
    return touch and f["middle"] and f["ring"] and f["pinky"]


def _is_thumb_pose_base(lm: List[Any], fingers: Optional[dict[str, bool]] = None) -> bool:
    f = fingers if fingers is not None else _all_fingers(lm)
    return (
        f["thumb"]
        and not f["index"]
        and not f["middle"]
        and not f["ring"]
        and not f["pinky"]
    )


def _is_thumbs_up(lm: List[Any], fingers: Optional[dict[str, bool]] = None) -> bool:
    if not _is_thumb_pose_base(lm, fingers):
        return False
    return lm[4].y < lm[3].y and lm[4].y < lm[0].y


def _is_thumbs_down(lm: List[Any], fingers: Optional[dict[str, bool]] = None) -> bool:
    if not _is_thumb_pose_base(lm, fingers):
        return False
    return lm[4].y > lm[3].y and lm[4].y > lm[0].y


def _is_lead_again(lm: List[Any], fingers: Optional[dict[str, bool]] = None) -> bool:
    """검지+엄지 ㄴ자 — 다시 리드해."""
    f = fingers if fingers is not None else _all_fingers(lm)
    if not (f["index"] and f["thumb"]):
        return False
    if f["middle"] or f["ring"] or f["pinky"]:
        return False
    touch = _dist(lm[4], lm[8]) < _dist(lm[5], lm[17]) * OK_TOUCH_RATIO
    return not touch


def _is_fist(lm: List[Any], fingers: Optional[dict[str, bool]] = None) -> bool:
    """주먹 — 나 따라와."""
    f = fingers if fingers is not None else _all_fingers(lm)
    return not any(f.values())


def classify_one_hand_gesture(lm: List[Any]) -> Optional[str]:
    fingers = _all_fingers(lm)
    if _is_open_palm(lm, fingers):
        return "stop"
    if _is_thumbs_up(lm, fingers):
        return "thumbs_up"
    if _is_thumbs_down(lm, fingers):
        return "thumbs_down"
    if _is_ok_sign(lm, fingers):
        return "ok_sign"
    if _is_lead_again(lm, fingers):
        return "lead_again"
    if _is_fist(lm, fingers):
        return "follow_me"
    return None
